File: leetcode_python/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head: return
        tail, cut, length = head, head, 0
        while tail.next:
            tail = tail.next
            length += 1
        k %= length + 1
        for i in range(length - k): cut = cut.next
        tail.next = head
        res, cut.next = cut.next, None
        return res

File: leetcode_java/61. Rotate List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null) return null;
        ListNode tail = head, cut = head;
        int len = 0;
        while(tail.next != null){
            tail = tail.next;
            len++;
        }
        k %= len + 1;
        for(int i = 0; i < len - k; i++) cut = cut.next;
        tail.next = head;
        ListNode res = cut.next;
        cut.next = null;
        return res;
    }
}